src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -609,4 +682,71 @@ public void cancelTimeout() {
             }
         }
     }
+
+    static class DirectResponseChannel implements TransportChannel {
+        final private String action;
+        final private long requestId;
+        final TransportServiceAdapter adapter;
+        final ThreadPool threadPool;
+
+        public DirectResponseChannel(String action, long requestId, TransportServiceAdapter adapter, ThreadPool threadPool) {
+            this.action = action;
+            this.requestId = requestId;
+            this.adapter = adapter;
+            this.threadPool = threadPool;
+        }
+
+        @Override
+        public String action() {
+            return action;
+        }
+
+        @Override
+        public void sendResponse(TransportResponse response) throws IOException {
+            sendResponse(response, TransportResponseOptions.EMPTY);
+        }
+
+        @Override
+        public void sendResponse(final TransportResponse response, TransportResponseOptions options) throws IOException {
+            final TransportResponseHandler handler = adapter.onResponseReceived(requestId);
+            // ignore if its null, the adapter logs it
+            if (handler != null) {
+                threadPool.executor(handler.executor()).execute(new Runnable() {
```

Comment:
I was doubting about this one. I ended op going for the simplest code, feeling it's not a big deal as most of the time this requests will go to another node. Please educate me if I'm wrong.